browser/brave_wallet/wallet_notification_service.cc

```diff
@@ -0,0 +1,98 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_wallet/wallet_notification_service.h"
+
+#include <memory>
+#include <string>
+
+#include "base/strings/utf_string_conversions.h"
+#include "brave/components/brave_wallet/browser/tx_service.h"
+#include "chrome/browser/notifications/notification_display_service.h"
+#include "chrome/browser/notifications/notification_display_service_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/base/models/image_model.h"
+#include "ui/message_center/public/cpp/notification.h"
+#include "ui/message_center/public/cpp/notification_types.h"
+#include "ui/message_center/public/cpp/notifier_id.h"
+
+namespace {
+
+std::unique_ptr<message_center::Notification> CreateMessageCenterNotification(
+    const std::u16string& title,
+    const std::u16string& body,
+    const std::string& uuid,
+    const GURL& link) {
+  message_center::RichNotificationData notification_data;
+  // hack to prevent origin from showing in the notification
+  notification_data.context_message = u" ";
+  auto notification = std::make_unique<message_center::Notification>(
+      message_center::NOTIFICATION_TYPE_SIMPLE, uuid, title, body,
+      ui::ImageModel(), std::u16string(), link,
+      message_center::NotifierId(message_center::NotifierType::SYSTEM_COMPONENT,
+                                 "service.wallet"),
+      notification_data, nullptr);
+
+  return notification;
+}
+
+void PushNotification(content::BrowserContext* context,
+                      const std::string& uuid,
+                      const std::string& from,
+                      const std::u16string& title,
+                      const std::u16string& body) {
+  auto notification = CreateMessageCenterNotification(
+      title, body, uuid,
+      GURL("brave://wallet/crypto/accounts/" + from + "#" + uuid));
+  auto* profile = Profile::FromBrowserContext(context);
+  NotificationDisplayServiceFactory::GetForProfile(profile)->Display(
+      NotificationHandler::Type::SEND_TAB_TO_SELF, *notification, nullptr);
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+WalletNotificationService::WalletNotificationService(
+    content::BrowserContext* context,
+    TxService* tx_service)
+    : tx_service_(tx_service), context_(context) {
+  tx_service_->AddObserver(tx_observer_receiver_.BindNewPipeAndPassRemote());
```

Comment:
This raw_ptr doesn't seem to be needed to be stored if all we need is the AddObserver call here.

Reply:
done